### PR TITLE
710 requisition repositories

### LINKED ISSUE #710: Requisition repositories
https://github.com/openmsupply/remote-server/issues/703

### PR DESCRIPTION
**Pull request by andreievg**
Closes #710 

### New repos

* Requisition
* RequistionRow
* RequistionLine
* RequisitionLineRow

In tests I tried testing filter query results against actual sql statement results.

It seems like sort and filters are better places in repositories, this avoids extra types (domain filter, sort, requisition_type and status). What do you think ?

### CODE REVIEW

File: repository/src/db_diesel/requisition/mod.rs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    pub r#type: Option<EqualFilter<RequisitionRowType>>,
+    pub status: Option<EqualFilter<RequisitionRowStatus>>,
+    pub created_datetime: Option<DatetimeFilter>,
+    pub sent_datetime: Option<DatetimeFilter>,
+    pub finalised_datetime: Option<DatetimeFilter>,
+    pub name_id: Option<EqualFilter<String>>,
+    pub name: Option<EqualFilter<String>>,
+    pub color: Option<EqualFilter<String>>,
+    pub their_reference: Option<EqualFilter<String>>,
+    pub comment: Option<EqualFilter<String>>,
```
Comment (clemens-msupply) [inline repository/src/db_diesel/requisition/mod.rs:24]: SimpleStringFilter might be a better fit for comment

File: repository/src/tests.rs [hunk 2]
```diff
@@ -896,42 +902,199 @@ mod repository_test {
...
-                .iter()
-                .enumerate()
-            {
-                assert_eq!(lines[count].id, line.id)
-            }
-        }
+            .unwrap();
+
+        let raw_result = sql_query(&format!(
+            r#"SELECT id from requisition where id = '{}'"#,
```
Comment (clemens-msupply) [inline repository/src/tests.rs:965]: In general this is sqlite/postgres specific isn't it, i.e. test might fail depending on the feature.

The repository is the high level interface to the database. However, using raw sql makes the test dependent on the implementation. The test should ensure that the interface is not broken, e.g. when changing the database schema for this reason it shouldn't use SQL. Better is to fill the repo with some test data and test that you get the correct data back (the only thing you really care about).

**Review by clemens-msupply — APPROVED** [review]
You mean move the common sort and filter, e.g. EqualFilter, to repository as well? Yes please having all of these defined where they originate seem to be more natural to me (and easier to implement as well since you just need to put it in your repository file). Think some repos already do it.

If we ever need something different in the services layer we can easily define some special filter/sorts there.

Please see comment regarding raw SQL in repo tests.

File: repository/src/db_diesel/requisition/mod.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            id: None,
+            requisition_number: None,
+            r#type: None,
+            status: None,
+            created_datetime: None,
+            sent_datetime: None,
+            finalised_datetime: None,
+            name_id: None,
+            name: None,
+            color: None,
```
[the remark this answers is not in the document]
Comment (andreievg) [reply repository/src/db_diesel/requisition/mod.rs]: thanks, 2a6aee7a9ac51e057164bc95e51579b54f8cf0bd

File: repository/src/db_diesel/requisition/mod.rs (hunk 1 above)
Comment (andreievg) [reply repository/src/db_diesel/requisition/mod.rs]: After talking to you and some thought, I agree, let's keep them seperate for now. 
also added sanity check (to make sure at least one record is matched).
11c39fc0be16056d59fe0b7cbd8ec5d0df41d9d2

File: repository/src/tests.rs (hunk 2 above)
Comment (andreievg) [reply repository/src/tests.rs:965]: I think if we are thinking that new repository implementation (api call etc..) are created then SQL test are not good, but if we are just testing the data access layer they seem pretty good to me (i think sql is pretty easy to read).

I'll try to remember to talk about it in our next meeting, hope it's ok to keep till then

**Pull request merged**